File: Sem-4/Algorithms/mergeSort.c
/*************************************************
Name : Shubham Dey
Program : Implement merge sort
Program No. : 
Date : 06.04.2022
**************************************************/

#include <stdio.h>
#include <stdlib.h>

int *memalloc(int n);
void input(int n, int *A);
void display(int n, int *A);
void mergeSort(int low, int up, int *A);
void merge(int low, int mid, int up, int *A);

int main()
{
	int *A,n;

	printf("How many numbers ? :\t");
	scanf("%d",&n);

	A = memalloc(n);

	printf("Enter %d elements :\n",n);
	input(n,A);
	printf("Unsorted Array :\n");
	display(n,A);

	mergeSort(0,n-1,A);

	printf("Sorted Array :\n");
	display(n,A);

	return 0;
}

int *memalloc(int n)
{
	int *A = (int*)calloc(n,sizeof(int));
	return A;
}

void input(int n, int *A)
{
	for(int i=0; i<n; i++)
		scanf("%d",&A[i]);
}

void display(int n, int *A)
{
	for(int i=0; i<n; i++)
		printf("%d\t",A[i]);
	printf("\n");
}

void mergeSort(int low, int up, int *A)
{
	if(low < up)
	{
		int mid = low + (up - low)/2;
		mergeSort(low, mid, A);
		mergeSort(mid+1, up, A);
		merge(low, mid, up, A);
	}
}

void merge(int low, int mid, int up, int *A)
{
	int i,j,k,n1,n2,*X,*Y;
	
	n1 = mid - low +1;
	n2 = up - mid;
	X = memalloc(n1);
	Y = memalloc(n2);

	for(i=0; i<n1; i++)
		X[i] = A[low+i];
	for(i=0; i<n2; i++)
		Y[i] = A[mid+1+i];

	i = j = 0;
	k = low;

	while(i<n1 && j<n2)
	{
		if(X[i]<=Y[j])
			A[k++] = X[i++];
		else
			A[k++] = Y[j++];
	}

	while(i<n1)
		A[k++] = X[i++];
	
	while(j<n2)
		A[k++] = Y[j++];

}

File: Sem-2/Java/executable/MergeSort.java
package executable;

import java.util.Scanner;

public class MergeSort {
	private Scanner sc = new Scanner(System.in);
	private void sort(int[] ar, int l, int u) {
		if(l<u) {
			int m = l + (u-l)/2;
			sort(ar, l, m);
			sort(ar, m+1, u);
			merge(ar, l, m, u);
		}
	}
	private void merge(int[] ar, int l, int m, int u) {
		int l1 = m-l+1;
		int l2 = u-m;
		int L[] = new int[l1];
		int R[] = new int[l2];
		
		for(int i=0; i<l1; i++)
			L[i] = ar[l+i];
		for(int i=0; i<l2; i++)
			R[i] = ar[m+i+1];
		
		int i=0, j=0, k=l;
		while(i<l1 && j<l2) {
			if(L[i]<=R[j]) {
				ar[k] = L[i];
				i++;
			}
			else {
				ar[k] = R[j];
				j++;
			}
			k++;
		}
		while(i<l1) {
			ar[k] = L[i];
			i++;
			k++;
		}
		while(j<l2) {
			ar[k] = R[j];
			j++;
			k++;
		}
	}
	private void display(int[] ar) {
		for(int i : ar)
			System.out.print(i + "  ");
		System.out.println();
	}
	private int[] input(int n) {
		int[] ar = new int[n];
		for(int i=0; i<n; i++)
			ar[i] = sc.nextInt();
		return ar;
	}
	public void run() {
		System.out.println("\nHow many numbers ? :\t");
		int n = sc.nextInt();
		MergeSort ms = new MergeSort();
		System.out.printf("Enter %d numbers :\n",n);
		int[] arr = ms.input(n);
		System.out.println("Given array :");
		ms.display(arr);
		ms.sort(arr, 0, arr.length-1);
		System.out.println("\nSorted array :");
		ms.display(arr);
		
		for(int i=0; i<30; i++)
			System.out.print("=");
		System.out.println();
	}
}